include/eve/detail/function/simd/x86/swap_adjacent_group.hpp

```diff
@@ -0,0 +1,183 @@
+//==================================================================================================
+/**
+  EVE - Expressive Vector Engine
+  Copyright : EVE Contributors & Maintainers
+  SPDX-License-Identifier: MIT
+**/
+//==================================================================================================
+#pragma once
+
+#include <eve/detail/abi.hpp>
+#include <eve/function/convert.hpp>
+#include <eve/pattern.hpp>
+
+namespace eve::detail
+{
+  template<real_scalar_value T, typename N, x86_abi ABI, std::ptrdiff_t G>
+  EVE_FORCEINLINE wide<T,N,ABI> swap_adjacent_group_( EVE_SUPPORTS(cpu_)
+                                                    , wide<T,N,ABI> v, fixed<G>
+                                                    ) noexcept
+  requires(G<=N::value)
+  {
+    using sf4     = std::integral_constant<int,_MM_SHUFFLE(2, 3, 0, 1)>;
+    using that_t  = wide<T,N,ABI>;
+
+    if constexpr(G == N::value)
+    {
+      return v;
+    }
+    else
+    {
+      constexpr auto size = G * sizeof(T);
+
+      if constexpr( std::same_as<ABI,x86_128_> )
+      {
+        if constexpr( size == 1 )
+        {
+          if constexpr( current_api >= ssse3 )
+          {
+            return _mm_shuffle_epi8 ( v.storage()
+                                    , _mm_set_epi8(14,15,12,13,10,11,8,9,6,7,4,5,2,3,0,1)
+                                    );
+          }
+          else
+          {
+            auto up = convert(v, as_<upgrade_t<T>>{});
+            auto r  = swap_adjacent_group(up,fixed<G>{});
+            auto[l,h] = r.slice();
+            if constexpr( std::is_signed_v<T> ) return _mm_packs_epi16(l,h);
+            else                                return _mm_packus_epi16(l,h);
+          }
+        }
+        else if constexpr( size == 2 )
+        {
+          auto s =v.storage();
+              s = _mm_shufflehi_epi16(s, sf4::value);
+          return  _mm_shufflelo_epi16(s, sf4::value);
+        }
+        else if constexpr( size == 4 )
+        {
+          if constexpr( std::same_as<T,float> ) return v[pattern<1,0,3,2>];
```

Comment:
Same thing as in the main - this won't fly.
And you do know the instruction, why not just call it?

Reply:
Will fix

Reply:
It also fails with smaller wide anyway